### PR TITLE
Plp integration - initial work

### PR DESCRIPTION
**Pull request by egillax**

### CODE REVIEW

File: cdm_processing/abstract_cdm_processor.py [hunk 1]
```diff
@@ -105,9 +105,12 @@ def _process_partition(self, partition_i: int):
...
                       available_tables}
-        if self._settings.label_sub_folder is not None:
-            labels = pq.read_table(os.path.join(self._settings.cdm_data_path,
-                                                self._settings.label_sub_folder, file_name))
+        if self._settings.labels is not None:
+            if self._settings.labels:
+                labels = pq.read_table(os.path.join(self._settings.cdm_data_path,
+                                                    "label", file_name))
+            else:
+                labels = pq.read_table(self._settings.labels)
```
Comment (egillax) [inline cdm_processing/abstract_cdm_processor.py]: To allow for `self._settings.labels` to be a path as well as a boolean

File: cdm_processing/cdm_processing_settings.py [hunk 2]
```diff
@@ -18,7 +18,7 @@ class CdmProcessingSettings:
     output_path: str
     mapping_settings: MappingSettings
     profile: bool
-    label_sub_folder: Optional[str] = None
+    labels: Optional[str] = None
```
Comment (egillax) [inline cdm_processing/cdm_processing_settings.py:21]: Renamed since it can now can be either a path or a boolean indicating a subfolder

File: cdm_processing/cdm_processor.py [hunk 3]
```diff
@@ -170,8 +170,12 @@ def _create_sequence_tables(self,
...
               "  ON visits.person_id = person.person_id"
         con.execute(sql)
-        if self._settings.label_sub_folder is not None:
-            part1 = "  label, "
+        if self._settings.labels is not None:
+            label_columns = con.query("SELECT column_name FROM information_schema.columns WHERE table_name = 'labels'")
+            if "is_training" in label_columns.arrow()["column_name"].to_pylist():
+                part1 = "  is_training,  label, "
+            else:
+                part1 = "  label, "
```
Comment (egillax) [inline cdm_processing/cdm_processor.py:184]: Added a new column to labels file, indicating whether the row is part of training or validation set. Necessary so  the PLP can control the split.

File: data_loading/dataset.py [hunk 4]
```diff
@@ -45,12 +48,40 @@ def __init__(self,
...
+                if not os.path.exists(test_dir):
+                    os.makedirs(test_dir)
+                out_files = []
+                for file in all_files:
+                    con = duckdb.connect()
+                    con.sql(f"CREATE VIEW sequence AS SELECT * FROM read_parquet('{file}')")
+                    output_path = os.path.join(test_dir, os.path.basename(file))
+                    # get test set and write to temp directory
+                    con.sql(f"COPY(SELECT * FROM sequence where is_training = false) TO '{output_path}' (FORMAT PARQUET)")
+                    out_files.append(output_path)
```
Comment (egillax) [inline data_loading/dataset.py:82]: split person_sequence into training and  validation set preserving partition structure.

File: model_trainer.yaml [hunk 5]
```diff
@@ -46,23 +51,23 @@ learning_objectives:
 
 
 training:
-  # Fraction of the data to use for training. Set to 1 to skip evaluation:
-  train_fraction: 0.8
+  # Fraction of the data to use for training. Set to 1 to skip evaluation or plp to use split from PLP in person sequence data:
+  train_fraction: plp
```
Comment (egillax) [inline model_trainer.yaml:55]: Maybe it's unnecessary to mention this option in the docs. It's only the PLP that can use it.

File: training/train_model.py [hunk 6]
```diff
@@ -294,11 +300,17 @@ def train_model(self) -> None:
...
 
     def evaluate_model(self, result_file: str, epoch: Optional[int] = None) -> None:
         if epoch is not None:
             self._load_model(file_name=os.path.join(self._settings.output_folder, _get_file_name(epoch)),
                              pretrained=True)
+        if self._settings.finetuned_epoch is not None:
+            self._load_model(file_name=os.path.join(self._settings.output_folder,
+                                                    _get_file_name(self._settings.finetuned_epoch)),
+                             pretrained=False)
         else:
```
Comment (egillax) [inline training/train_model.py:316]: I want the PLP to be able to select model from epoch based on best validation performance. Common for models that are very flexible and can easily overfit such as deep neural networks and gradient boosted trees.

File: utils/results.py [hunk 7]
```diff
@@ -20,4 +21,28 @@ def write_to_csv(self, file_name: str):
...
+            self.metrics_list.append({"epochs": epochs, "metrics": metrics})
+        else:
+            self.metrics_list.append({"epochs": epochs, "metrics": metrics})
+
+    def flush(self):
+        if self._file_name:
+            with open(self._file_name, "a", encoding="UTF8", newline="") as f:
+                f.write(json.dumps(self.metrics_list, indent=4))
+        else:
+            raise ValueError("No file name specified for json writer")
```
Comment (egillax) [inline utils/results.py]: I needed some easily parse-able data for the PLP to use to select the best epoch. Maybe it was stubborness in me but I thought json was better for this than the already present Results writer.
